[2기-C 홍유석] JPA 게시판 구현

yshjft opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
JPA 게시판 구현 과제 입니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

### 요구 사항 및 구현 여부
- [X] SpringDataJPA를 설정한다.
   - datasource: h2
- [X] 엔티티를 구성한다.
- [X] API를 구현한다.
    - [X] 게시글 조회
       - [X] 페이징 조회 (GET "/posts")
       - [X] 단건 조회 (GET "/posts/{id}")
    - [X] 게시글 작성 (POST "/posts")
    - [X] 게시글 수정 (POST "/posts/{id}")
- [X] REST-DOCS를 이용해서 문서화한다.

### api 상세
api에 대한 자세한 내용은 [REST-DOCS를 이용하여 문서화한 것](https://yshjft.github.io/jpa-board-api-docs/)으로 대체하겠습니다.

yshjft committed:
feat: init project

yshjft committed:
feat: User, Post 엔티티 생성

yshjft committed:
feat: UserRepository와 이에 대한 test code 작성 완료

yshjft committed:
feat:  PostRepository와 이에 대한 테스트 코드 작성 완료

yshjft committed:
feat: 회원 가입 api 구현

yshjft committed:
feat: 예외 처리 구현

yshjft committed:
feat: 게시물 생성 api 구현 및 UserNotFound 예외 처리

yshjft committed:
fix: 게시물 생성 service layer 코드 수정

yshjft committed:
feat: 게시글 페이징 조회 구현

yshjft committed:
feat: 게시물 조회 구현

yshjft committed:
feat: 게시글 수정 구현

yshjft committed:
feat: 문서화 추가

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/controller/PostController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<ResponseDto> updatePost(@PathVariable Long id, @RequestBody PostUpdateDto postUpdateDto) {
```
PostUpdateDto에 @Valid 달아주시는거 잊으신 것 같아요!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,107 @@
+    private User user;
+
+    public Post() {
+    }
```
이미 알고계시는거 알지만 나중에 리팩토링 편하게 하시라고 코멘트 남깁니다!
- 엔티티 기본 생성자의 접근 제한자 수정하기

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/user/domain/User.java:
```diff
@@ -0,0 +1,111 @@
+    private List<Post> posts = new ArrayList<>();
+
+    public User() {
+    }
```
이미 알고계시는거 알지만 나중에 리팩토링 편하게 하시라고 코멘트 남깁니다!
- 엔티티 기본 생성자의 접근 제한자 수정하기

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,107 @@
+
+    public Post(String createdBy, LocalDateTime createdAt, LocalDateTime updatedAt, Long id, String title, String content) {
+        super(createdBy, createdAt, updatedAt);
+        this.id = id;
```
PostService 테스트 때문에 id가 들어간 것 같은데 맞나요??
id값은 @GeneratedValue로 생성되는 값이기 때문에 Post 생성자랑 빌더에서 값을 넣어주지 못하게 빼야된다고 생각합니다!
테스트는 다른 방식을 생각해봐야할 것 같아요!ㅠㅠ

yshjft replied:
생성자를 단축키로 생성하다보니 아무 생각 없이 id를 상용하고 있었네요;; 수정하겠습니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/global/common/response/ResultDto.java:
```diff
@@ -0,0 +1,34 @@
+import java.time.LocalDateTime;
+
+@Getter
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
이 애노테이션 사용해서 상황에 따라 필요한 값만 들어가게 하는거 좋은 것 같아요!👍
또 하나 배우고 갑니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/global/common/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+    protected LocalDateTime updatedAt;
+
+    public BaseEntity() {
+    }
```
이미 알고계시는거 알지만 나중에 리팩토링 편하게 하시라고 코멘트 남깁니다!
- 엔티티 기본 생성자의 접근 제한자 수정하기

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/user/dto/UserRequestDto.java:
```diff
@@ -0,0 +1,41 @@
+    private String name;
+
+    @Min(1)
+    private int age;
```
User Entity를 보니 age는 nullable한 값이기 때문에 null을 표현할 수 있도록 primitive 보단 wrapper를 사용하면 좋을 것 같습니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/user/domain/User.java:
```diff
@@ -0,0 +1,111 @@
+
+    private int age;
+
+    private String hobby;
```
UserRequestDto를 보면 hobby의 최대 length를 255로 해두셨는데,
hobby column에 length 제약조건을 추가하면 어떨까요?

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/user/domain/User.java:
```diff
@@ -0,0 +1,111 @@
+    @Column(length = 50, nullable = false)
+    private String name;
+
+    private int age;
```
age는 nullable한 값이기 때문에 null을 표현할 수 있도록 primitive 보단 wrapper를 사용하면 좋을 것 같습니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,90 @@
+        return ResultDto.createResult(savedPost.getId());
+    }
+
+    @Transactional(readOnly = true)
```
트랜잭션을 읽기 전용으로 해두면 엔티티의 등록, 변경, 삭제를 막고 성능까지 최적화할 수 있군요!
잘 모르고 있던 옵션인데 유석님 덕분에 배웠습니다!!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@Slf4j
```
사용하지 않으니 지워도 될 것 같습니다!!

HyeonseoJUNG commented on src/test/java/com/prgrms/jpaboard/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class PostServiceTest {
```
게시글 단건 조회, 페이징 조회, 업데이트 테스트도 추가하면 좋을 것 같습니다:)

yshjft replied:
게시글 단건 조회와 패이지 조회의 경우 단순 조회와 조회된 데이터를 dto로 변경만 하는 로직뿐이라 테스트 코드 작성을 생략하였습니다.  업데이트의 경우 entity의 field 업데이트 메서드가 잘 동작하는지에 대한 테스트 코드를 추가하도록 하겠습니다.

HyeonseoJUNG commented on src/test/java/com/prgrms/jpaboard/domain/post/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,149 @@
+@AutoConfigureRestDocs
+@AutoConfigureMockMvc
+@SpringBootTest
+class PostControllerTest {
```
DB에 들어있는 값에 따라 테스트 성공 여부가 달라지는데,
DB에 들어있는 데이터에도 상관없이 테스트가 성공할 수 있도록 바꾸면 좋을 것 같습니다!

HyeonseoJUNG reviewed (commented):
안녕하세요! 유석님😊
게시판 과제 너무 고생하셨습니다!
코드리뷰 완료했습니다~</br>
DTO를 굉장히 세분화해서 잘 작성해주셔서 보면서 많이 배웠습니다👍
그리고 예외처리하는 부분에서도 많이 배웠어요!!
감사합니다😄

</br></br>

추가적으로
1. lombok으로 getter랑 constructor 만들 때만 사용하신 것 같은데, lombok을 사용하지 않고 해봐도 좋을 것 같아요!
2. 데이터 유효성 검증이 entity에는 없는데 dto에서 하더라도 entity에서도 해주는게 좋을 것 같습니다!

HyeonseoJUNG commented on src/main/java/com/prgrms/jpaboard/global/common/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+
+@Getter
+@MappedSuperclass
+public class BaseEntity {
+    @Column(name="created_by", nullable = false)
+    private String createdBy;
+
+    @Column(name="created_at", columnDefinition = "TIMESTAMP", updatable = false, nullable = false)
+    private LocalDateTime createdAt;
+
+    @Column(name="updated_at", columnDefinition = "TIMESTAMP", nullable = false)
+    protected LocalDateTime updatedAt;
```
1. BaseEntity를 추상 클래스로 만들기
2. Entity Listener를 통해 JPA Auditing 사용하기

[참고: BaseEntity & JPA Auditing](https://wildeveloperetrain.tistory.com/76)
> @MappedSupperclass가 선언된 클래스는 Entity가 아니며 직접 생성해서 사용될 일이 없기 때문에 대부분 추상 클래스로 만들어집니다.

[참고: Entity Listener](https://milenote.tistory.com/79)

HyeonseoJUNG reviewed (commented):
안녕하세요! 유석님:)
오늘 스크럼 시간에 말씀드렸던 부분 추가적으로 코멘트 달았습니다😊

yshjft committed:
fix: 엔티티 수정

(1) 기본 생성자의 접근 제한자를 protected로 변경

(2) 엔티티의 생성자에서 id 설정하지 않도록 수정

yshjft committed:
fix: 테스트 코드 수정 및 추가, dto와 엔티티 수정

(1) 테스트 코드 수정 및 추가
게시글 업데이트 테스트 추가

(2) dto와 엔티티 수정
age가 DB에서 nullable=true이므로 dto와 엔티티에서 해당 값의 타입을 int에서 Integer로 변경

yshjft committed:
fix: controller 테스트 코드 수정, dto 수정

(1) 테스트 코드 수정
controller 테스트 코드 DB에 영향을 받지 않도록 @SpringBootTest에서 @WebMvcTest로 수정

(2) dto 수정
- 페이징 정보를 담던 클래스의 이름을 MetaDataDto에서 PagingInfo로 변경
- 테스트 코드에서의 사용을 위해 일부 DTO에 equals and hashcode 추가

yshjft committed:
fix: jpa auditing 적용과 그에 따른 dto 변경

jpa auditing을 적용하였으며 ResultDto에서 createdAt과 updatedAt을 제거 하였습니다.

yshjft committed:
fix: getter 수정

사용하는 getter만 존재하도록 엔티티 클래스에서 @Getter를 제거하였습니다. 또한 각각의 엔티티에서 DTO를 반환할 수 있도록 메서드를 구현하였습니다.

yshjft committed:
fix: service와 entity에서 validation 추가
